evm-ds/src/main.rs

```diff
@@ -98,44 +93,123 @@ impl CallContext {
     }
 }
 
+#[derive(Debug,Serialize,Deserialize)]
+struct StructLog {
+    pub depth: usize,
+    pub error: String,
+    pub gas: u64, // not populated
+    #[serde(rename = "gasCost")]
+    pub gas_cost: u64, // not populated
+    pub op: String,
+    pub pc: usize,
+    pub stack: Vec<String>,
+    pub storage: Vec<String>, // not populated
+}
+
+impl StructLog {
+    fn new() -> Self{
+        StructLog{
+            depth: Default::default(),
+            error: Default::default(),
+            gas: Default::default(),
+            gas_cost: Default::default(),
+            op: Default::default(),
+            pc: Default::default(),
+            stack: Default::default(),
+            storage: Default::default(),
+        }
+    }
+}
 
 // This implementation has a stack of call contexts each with reference to their calls - so a tree is
 // Created in this way.
 // Each new call gets added to the end of the stack and becomes the current context.
 // On returning from a call, the end of the stack gets put into the item above's calls
 #[derive(Debug,Serialize,Deserialize)]
 struct LoggingEventListener {
-    call_stack: Vec<CallContext>,
+    call_tracer: Vec<CallContext>,
+    raw_tracer: StructLogTopLevel,
     enabled: bool,
 }
 
-impl LoggingEventListener {
+#[derive(Debug,Serialize,Deserialize, Default)]
+struct StructLogTopLevel {
+    pub gas: u64,
+    #[serde(rename = "returnValue")]
+    pub return_value: String,
+    #[serde(rename = "structLogs")]
+    pub struct_logs: Vec<StructLog>,
+}
 
+impl LoggingEventListener {
     fn new(enabled: bool) -> Self {
         LoggingEventListener {
-            call_stack: Default::default(),
+            call_tracer: Default::default(),
+            raw_tracer: Default::default(),
             enabled: enabled,
         }
     }
 }
 
-impl tracing::EventListener for LoggingEventListener {
-    fn event(&mut self, event: tracing::Event) {
-        println!("recvd event: {:?}", event);
+impl evm::runtime::tracing::EventListener for LoggingEventListener {
+    fn event(&mut self, event: evm::runtime::tracing::Event) {
+
+        if !self.enabled {
+            return;
+        }
+
+        let mut struct_log = StructLog::new();
```

Comment:
```suggestion
        let mut struct_log = StructLog { depth: self.call_tracer.len() - 1, ..Default::default() };
```

Reply:
Great, I didn't know you could do such patterns :)